### PR TITLE
Update the code of service y2r!

### PR DESCRIPTION
**Pull request by JamePeng**
I rebase my fork code and redo it ,and I give up the alignment for compatibility
This change will fixed the Application EX Troopers(エクストルーパーズ) some Graphics problem！

### CODE REVIEW

File: src/core/hle/service/y2r_u.cpp [hunk 1]
```diff
@@ -374,50 +483,50 @@ static void DriverFinalize(Service::Interface* self) {
...
+    {0x00180102, SetReceiving,                "SetReceiving"},
+    {0x00190000, nullptr,                     "IsFinishedReceiving"},
+    {0x001A0040, SetInputLineWidth,           "SetInputLineWidth"},
+    {0x001B0000, GetInputLineWidth,           "GetInputLineWidth"},
+    {0x001C0040, SetInputLines,               "SetInputLines"},
+    {0x001D0000, GetInputLines,               "GetInputLines"},
+    {0x001E0100, SetCoefficient,              "SetCoefficient"},
+    {0x001F0000, GetCoefficient,              "GetCoefficient"},
+    {0x00200040, SetStandardCoefficient,      "SetStandardCoefficient"},
+    {0x00210040, GetStandardCoefficientParams,"GetStandardCoefficientParams"},
```
Comment (yuriks) [inline src/core/hle/service/y2r_u.cpp]: MIssing space after `,`

File: src/core/hle/service/y2r_u.h [hunk 2]
```diff
@@ -97,6 +97,7 @@ struct ConversionConfiguration {
     u16 input_line_width;
     u16 input_lines;
     CoefficientSet coefficients;
+    u8  padding;
```
Comment (yuriks) [inline src/core/hle/service/y2r_u.h]: Don't align the variable name.

File: src/core/hle/service/y2r_u.cpp [hunk 3]
```diff
@@ -332,6 +433,14 @@ static void SetConversionParams(Service::Interface* self) {
     cmd_buff[1] = result.raw;
 }
 
+static void GetPackageParameter(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    std::memcpy(&cmd_buff[2], reinterpret_cast<const u32*>(&(conversion)), sizeof(ConversionParameters));
```
Comment (yuriks) [inline src/core/hle/service/y2r_u.cpp]: The reinterpret_cast is pointless.

File: src/core/hle/service/y2r_u.cpp [hunk 4]
```diff
@@ -259,6 +350,15 @@ static void SetAlpha(Service::Interface* self) {
     cmd_buff[1] = RESULT_SUCCESS.raw;
 }
 
+static void GetAlpha(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    LOG_DEBUG(Service_Y2R, "called Get alpha=%hu", conversion.alpha);
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+    cmd_buff[2] = static_cast<u32>(conversion.alpha);
```
Comment (yuriks) [inline src/core/hle/service/y2r_u.cpp]: This cast isn't necessary.

File: src/core/hle/service/y2r_u.cpp (hunk 3 above)
Comment (yuriks) [inline src/core/hle/service/y2r_u.cpp]: As are the parenthesis around `conversion`.

File: src/core/hle/service/y2r_u.cpp [hunk 5]
```diff
@@ -250,6 +312,35 @@ static void SetStandardCoefficient(Service::Interface* self) {
     cmd_buff[1] = conversion.SetStandardCoefficient((StandardCoefficient)cmd_buff[1]).raw;
 }
 
+static void GetStandardCoefficientParams(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    u16  index = 4;
```
Comment (yuriks) [inline src/core/hle/service/y2r_u.cpp]: Why is this a u16??? And why does it default to 4? And don't try to column align variables.

File: src/core/hle/service/y2r_u.cpp [hunk 6]
```diff
@@ -242,6 +296,14 @@ static void SetCoefficient(Service::Interface* self) {
     cmd_buff[1] = RESULT_SUCCESS.raw;
 }
 
+static void GetCoefficient(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    std::memcpy(&cmd_buff[2], conversion.coefficients.data(), sizeof(CoefficientSet));
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
```
Comment (yuriks) [inline src/core/hle/service/y2r_u.cpp]: Try to assign cmd_buff elements in ascending order.

File: src/core/hle/service/y2r_u.cpp (hunk 3 above)
Comment (JamePeng) [inline src/core/hle/service/y2r_u.cpp]: @yuriks 
(const u32*)&(conversion) ?

Comment (mailwl) [inline src/core/hle/service/y2r_u.cpp]: just conversion
and instead memcpy, but Memory::WriteBlock

File: src/core/hle/service/y2r_u.cpp (hunk 5 above)
Comment (JamePeng) [inline src/core/hle/service/y2r_u.cpp]: @yuriks Because the index only 0 to 3, if index nothing change,the 4 is invalid! 
it make sure the cmd_buff[1] input a invalid value, and the if-check can work normally

File: src/core/hle/service/y2r_u.cpp (hunk 6 above)
Comment (JamePeng) [inline src/core/hle/service/y2r_u.cpp]: @yuriks hmm? what do you mean?

Comment (yuriks) [inline src/core/hle/service/y2r_u.cpp]: You're assigning to cmd_buff[2] and up, and then to cmd_buff[1], it's kinda confusing.

File: src/core/hle/service/y2r_u.cpp [hunk 7]
```diff
@@ -250,6 +312,35 @@ static void SetStandardCoefficient(Service::Interface* self) {
     cmd_buff[1] = conversion.SetStandardCoefficient((StandardCoefficient)cmd_buff[1]).raw;
 }
 
+static void GetStandardCoefficientParams(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    u32 index = 4;
```
Comment (yuriks) [inline src/core/hle/service/y2r_u.cpp]: At least use "-1" or something more obviously special instead then.

File: src/core/hle/service/y2r_u.cpp [hunk 8]
```diff
@@ -250,6 +312,35 @@ static void SetStandardCoefficient(Service::Interface* self) {
     cmd_buff[1] = conversion.SetStandardCoefficient((StandardCoefficient)cmd_buff[1]).raw;
 }
 
+static void GetStandardCoefficientParams(Service::Interface* self) {
```
Comment (mailwl) [inline src/core/hle/service/y2r_u.cpp]: all this function may looks like:
```
static void GetStandardCoefficientParams(Service::Interface* self) {
    u32* cmd_buff = Kernel::GetCommandBuffer();
    u32 index = cmd_buff[1];
    if(index < 4) { // instead magic number 4 cast to enum maybe
        std::memcpy(&cmd_buff[2], &standard_coefficients[index], sizeof(CoefficientSet));
        LOG_WARNING(Service_Y2R, "StandardCoefficient:0x%08X ", index);
        cmd_buff[1] = RESULT_SUCCESS.raw;
    } else {
        LOG_ERROR(Service_Y2R, "StandardCoefficient:0x%08X  The argument is invalid!", index);
        cmd_buff[1] = -1;
    }
}
```

Comment (JamePeng) [inline src/core/hle/service/y2r_u.cpp]: @mailwl I am agree!

File: src/core/hle/service/y2r_u.cpp [hunk 9]
```diff
@@ -301,17 +387,17 @@ static void IsBusyConversion(Service::Interface* self) {
 /**
  * Y2R_U::SetConversionParams service function
```
Comment (wwylele) [inline src/core/hle/service/y2r_u.cpp]: Update this comment since you change the function name

File: src/core/hle/service/y2r_u.cpp [hunk 10]
```diff
@@ -332,6 +411,15 @@ static void SetConversionParams(Service::Interface* self) {
     cmd_buff[1] = result.raw;
 }
 
+static void GetPackageParameter(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+    std::memcpy(&cmd_buff[2], &(conversion), sizeof(ConversionParameters));
```
Comment (wwylele) [inline src/core/hle/service/y2r_u.cpp]: `&(conversion)`->`&conversion`

**Comment by JamePeng** [comment]
Ready for merge!

**Comment by yuriks** [comment]
Can you split the changes here that are only formatting, and the ones that are actually functional? I don't really mind the formatting changes, but it gets hard to review since I'm not sure what actually changed because of the diff churn. Also a better commit message would be good.

**Comment by JamePeng** [comment]
Yes, this commit was done after " rebase -i ", I have to redo it now!

**Comment by JamePeng** [comment]
As yuriks says,this is functional commit.
If the code is fine, I will align the code soon!

File: src/core/hle/service/y2r_u.cpp [hunk 11]
```diff
@@ -91,6 +96,14 @@ static void SetInputFormat(Service::Interface* self) {
     cmd_buff[1] = RESULT_SUCCESS.raw;
 }
 
+static void GetInputFormat(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+    cmd_buff[2] = static_cast<u32>(conversion.input_format);
+    LOG_DEBUG(Service_Y2R, "Get input_format=%hhu", conversion.input_format);
```
Comment (bunnei) [inline src/core/hle/service/y2r_u.cpp]: Same comment as previous PRs on your inconsistent logging... Get rid of the "Get ", this can just be:
<pre>
LOG_DEBUG(Service_Y2R, "input_format=%hhu", conversion.input_format);
</pre>

File: src/core/hle/service/y2r_u.cpp [hunk 12]
```diff
@@ -100,6 +113,14 @@ static void SetOutputFormat(Service::Interface* self) {
     cmd_buff[1] = RESULT_SUCCESS.raw;
 }
 
+static void GetOutputFormat(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+    cmd_buff[2] = static_cast<u32>(conversion.output_format);
+    LOG_DEBUG(Service_Y2R, "Get output_format=%hhu", conversion.output_format);
```
Comment (bunnei) [inline src/core/hle/service/y2r_u.cpp]: Same here... I'll stop pointing it out

File: src/core/hle/service/y2r_u.cpp [hunk 13]
```diff
@@ -118,15 +147,104 @@ static void SetBlockAlignment(Service::Interface* self) {
     cmd_buff[1] = RESULT_SUCCESS.raw;
 }
 
+static void GetBlockAlignment(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+    cmd_buff[2] = static_cast<u32>(conversion.block_alignment);
+    LOG_DEBUG(Service_Y2R, "called Get block_alignment=%hhu", conversion.block_alignment);
```
Comment (bunnei) [inline src/core/hle/service/y2r_u.cpp]: Especially here... "called Get ..." ? That's just goofy :)

File: src/core/hle/service/y2r_u.cpp [hunk 14]
```diff
@@ -118,15 +147,104 @@ static void SetBlockAlignment(Service::Interface* self) {
...
+ *      1 : u8, 0 = Disabled, 1 = Enabled
+ *  Outputs:
+ *      1 : Result of function, 0 on success, otherwise error code
+ */
+static void SetSpacialDithering(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+    is_enable_spacial_dithering = cmd_buff[1];
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+    LOG_DEBUG(Service_Y2R, "(STUBBED) called");
```
Comment (bunnei) [inline src/core/hle/service/y2r_u.cpp]: Any stubbed function should LOG_WARNING

File: src/core/hle/service/y2r_u.cpp [hunk 15]
```diff
@@ -118,15 +147,104 @@ static void SetBlockAlignment(Service::Interface* self) {
...
+ *  Outputs:
+ *      1 : Result of function, 0 on success, otherwise error code
+ *      2 : u8, 0 = Disabled, 1 = Enabled
+ */
+static void GetSpacialDithering(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+    cmd_buff[2] = is_enable_spacial_dithering;
+    LOG_DEBUG(Service_Y2R, "(STUBBED) called");
```
Comment (bunnei) [inline src/core/hle/service/y2r_u.cpp]: LOG_WARNING

File: src/core/hle/service/y2r_u.cpp [hunk 16]
```diff
@@ -118,15 +147,104 @@ static void SetBlockAlignment(Service::Interface* self) {
...
+ *      1 : u8, 0 = Disabled, 1 = Enabled
+ *  Outputs:
+ *      1 : Result of function, 0 on success, otherwise error code
+ */
+static void SetTemporalDithering(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+    is_enable_temporal_dithering = cmd_buff[1];
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+    LOG_DEBUG(Service_Y2R, "(STUBBED) called");
```
Comment (bunnei) [inline src/core/hle/service/y2r_u.cpp]: LOG_WARNING

File: src/core/hle/service/y2r_u.cpp [hunk 17]
```diff
@@ -118,15 +147,104 @@ static void SetBlockAlignment(Service::Interface* self) {
...
+ *  Outputs:
+ *      1 : Result of function, 0 on success, otherwise error code
+ *      2 : u8, 0 = Disabled, 1 = Enabled
+ */
+static void GetTemporalDithering(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+    cmd_buff[2] = is_enable_temporal_dithering;
+    LOG_DEBUG(Service_Y2R, "(STUBBED) called");
```
Comment (bunnei) [inline src/core/hle/service/y2r_u.cpp]: LOG_WARNING ... I'll stop pointing these out

File: src/core/hle/service/y2r_u.cpp [hunk 18]
```diff
@@ -200,6 +318,38 @@ static void SetSendingYUYV(Service::Interface* self) {
     cmd_buff[1] = RESULT_SUCCESS.raw;
 }
 
+static void IsFinishedSendingYuv(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    cmd_buff[1] = RESULT_SUCCESS.raw;
+    cmd_buff[2] = 1;
```
Comment (bunnei) [inline src/core/hle/service/y2r_u.cpp]: Add a comment saying what this is... "Is finished sending YUV" ?

File: src/core/hle/service/y2r_u.cpp [hunk 19]
```diff
@@ -200,6 +318,38 @@ static void SetSendingYUYV(Service::Interface* self) {
...
+    LOG_DEBUG(Service_Y2R, "(STUBBED) called IsFinishedSendingYuv=%u", cmd_buff[2]);
```
Comment (bunnei) [inline src/core/hle/service/y2r_u.cpp]: I don't see a point in logging a value you're setting (1)... get rid of it, and also change these to LOG_WARNING of course

Comment (bunnei) [inline src/core/hle/service/y2r_u.cpp]: Ditto on the duplication below

File: src/core/hle/service/y2r_u.cpp [hunk 20]
```diff
@@ -28,13 +29,17 @@ struct ConversionParameters {
...
-    u8 reserved;
+    u8 padding;
     u16 alpha;
 };
 static_assert(sizeof(ConversionParameters) == 12, "ConversionParameters struct has incorrect size");
 
 static Kernel::SharedPtr<Kernel::Event> completion_event;
 static ConversionConfiguration conversion;
+static DitheringWeightParams dithering_weight_params;
+static u32 is_enable_temporal_dithering = 0;
```
Comment (Subv) [inline src/core/hle/service/y2r_u.cpp]: temporal_dithering_enabled

File: src/core/hle/service/y2r_u.cpp [hunk 21]
```diff
@@ -28,13 +29,17 @@ struct ConversionParameters {
...
+static u32 is_enable_transfer_end_interrupt = 0;
```
Comment (Subv) [inline src/core/hle/service/y2r_u.cpp]: transfer_end_interrupt_enabled

File: src/core/hle/service/y2r_u.cpp [hunk 22]
```diff
@@ -28,13 +29,17 @@ struct ConversionParameters {
...
+static u32 is_enable_spacial_dithering = 0;
```
Comment (Subv) [inline src/core/hle/service/y2r_u.cpp]: spacial_dithering_enabled

File: src/core/hle/service/y2r_u.cpp [hunk 23]
```diff
@@ -109,21 +130,116 @@ static void SetRotation(Service::Interface* self) {
...
+/**
+ * Y2R_U::GetTransferEndInterrupt service function
+ *  Outputs:
+ *      1 : Result of function, 0 on success, otherwise error code
+ *      2 : u8, 0 = Disabled, 1 = Enabled
+ */
+static void GetTransferEndInterrupt(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
 
-    cmd_buff[0] = IPC::MakeHeader(0xD, 1, 0);
```
Comment (Subv) [inline src/core/hle/service/y2r_u.cpp]: Why did you remove cmd_buff[0] from this function?

File: src/core/hle/service/y2r_u.cpp [hunk 24]
```diff
@@ -279,7 +530,6 @@ static void StartConversion(Service::Interface* self) {
 static void StopConversion(Service::Interface* self) {
     u32* cmd_buff = Kernel::GetCommandBuffer();
 
-    cmd_buff[0] = IPC::MakeHeader(0x27, 1, 0);
```
Comment (Subv) [inline src/core/hle/service/y2r_u.cpp]: Why?

File: src/core/hle/service/y2r_u.cpp [hunk 25]
```diff
@@ -325,13 +575,22 @@ static void SetConversionParams(Service::Interface* self) {
     if (result.IsError()) goto cleanup;
     result = conversion.SetStandardCoefficient(params->standard_coefficient);
     if (result.IsError()) goto cleanup;
+    conversion.padding = params->padding;
     conversion.alpha = params->alpha;
 
 cleanup:
-    cmd_buff[0] = IPC::MakeHeader(0x29, 1, 0);
```
Comment (Subv) [inline src/core/hle/service/y2r_u.cpp]: Why?

Comment (JamePeng) [inline src/core/hle/service/y2r_u.cpp]: Because the function_table was register the function before, there are no need to register again!

Comment (wwylele) [inline src/core/hle/service/y2r_u.cpp]: no, this line is response header, which is different with the one in function table

Comment (wwylele) [inline src/core/hle/service/y2r_u.cpp]: actually all service function should give such a response header. However games usually don't check it, and most of them are missing in citra.

Comment (JamePeng) [inline src/core/hle/service/y2r_u.cpp]: @wwylele 
Actually it is not necessary, because the cmd_buff[0] is a copy which from getcommandbuffer!
there are no need to do it again!

Comment (wwylele) [inline src/core/hle/service/y2r_u.cpp]: @JamePeng (from what I understand) the header ('cmd_buf[0]') describes how many parameters following. Game process write request parameters to cmd_buf, including the header, and send it to service process. Then the service process write response back to the cmd_buf, with different header since response parameters are usually different from the request one.

Comment (JamePeng) [inline src/core/hle/service/y2r_u.cpp]: @wwylele  
My point is from libctru?
qq 20160420171019
The cmd_buff[0] is no how many parameters

Comment (wwylele) [inline src/core/hle/service/y2r_u.cpp]: It is. the '2' in IPC_MakeHeader indicates there are 2 words as normal parameter and 0 indicates there is no translated parameter. See further infomation on [this page](https://www.3dbrew.org/wiki/IPC)

Comment (JamePeng) [inline src/core/hle/service/y2r_u.cpp]: ok!

**Comment by JamePeng** [comment]
I append the IPC::MakeHeader to function!

**Comment by bunnei** [comment]
Looks fine
